### PR TITLE
add `mackerel-plugin` command

### PR DESCRIPTION
**Pull request by Songmu**
- Bundle the plugins to single binary (`mackerel-plugin`)
- Dispatch plugin processing with argument or symbolic link naming
  - eg.
  - `mackerel-plugin memcached` runs memcached plugin.
  - When a symbolic link named `mackerel-plugin-memcached` was linked to `mackerel-plugin`, run the memcached plugin
- There are no breaking changes.

### CODE REVIEW

File: tool/gen_mackerel_plugin.pl [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+func runPlugin(plug string) error {
+	switch plug {
+${case}\tdefault:
+		return fmt.Errorf("unknown plugin: %q", plug)
+	}
+	return nil
+}
+
+var plugins = []string{
```
Comment (haya14busa) [inline tool/gen_mackerel_plugin.pl]: [nit] You can use `[...]string{...}` to make plugins var array, not slice.

https://golang.org/ref/spec#Composite_literals

File: mackerel-plugin.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+func printHelp() {
+	fmt.Printf(`mackerel-plugin %s
+
+Usage: mackerel-plugin <plugin> [<args>]
+
+Following plugins are available:
+    %s
+
+See `+"`mackerel-plugin <plugin> -h` "+`for more information on a specific plugin
```
Comment (astj) [inline mackerel-plugin.go]: [nit]
filename of this binary (maybe `mackerel-plugin`) will be shown in each plugin's usage on called `mackerel-plugin <plugin> -h`.
Do you have any idea to solve this? (I think it's difficult…)
```
[github.com/mackerelio/mackerel-agent-plugins/build]$ ./mackerel-agent-plugins uptime -help
Usage of ./mackerel-agent-plugins:
  -metric-key-prefix string
    	Metric key prefix (default "uptime")
  -tempfile string
    	Temp file name
[github.com/mackerelio/mackerel-agent-plugins/build]$ ls -la ./mackerel-plugin-uptime
lrwxr-xr-x  1 astj  staff  22  3  6 16:12 ./mackerel-plugin-uptime@ -> mackerel-agent-plugins
[github.com/mackerelio/mackerel-agent-plugins/build]$ ./mackerel-plugin-uptime -help
Usage of ./mackerel-plugin-uptime:
  -metric-key-prefix string
    	Metric key prefix (default "uptime")
  -tempfile string
    	Temp file name
```

File: tool/gen_mackerel_plugin.pl [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env perl
+use 5.014;
+use warnings;
+use utf8;
+use autodie;
+use JSON::PP;
+
+my $json = do {
+    local $/;
+    open my $fh, '<', 'packaging/config.json';
```
Comment (astj) [inline tool/gen_mackerel_plugin.pl]: 👍

File: tool/gen_mackerel_plugin.pl [hunk 4]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env perl
```
Comment (astj) [inline tool/gen_mackerel_plugin.pl]: Do you have any idea when to run this script in our development flow?  Apparently we must run it when packaging/config.json changed, but one might forget to run.  IMO we need some automated process in somewhere (ci, release process, or something).

File: tool/gen_mackerel_plugin.pl [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+my $json = do {
+    local $/;
+    open my $fh, '<', 'packaging/config.json';
+    <$fh>
+};
+
+my @plugins = sort @{ decode_json($json)->{plugins}};
+
+sub resolve_package {
```
Comment (astj) [inline tool/gen_mackerel_plugin.pl]: Currently there's no guarantee that all plugin's package is named in this way.
Should we add some documentations for contributors?

File: tool/gen_mackerel_plugin.pl (hunk 4 above)
Comment (Songmu) [inline tool/gen_mackerel_plugin.pl]: I added test code to Makefile to detect missing `go generate` execution as following.


```shell
test `go generate ./... && git diff | wc -l` = 0 || (echo 'please `go generate ./...` and commit them' && exit 1)
```

File: tool/gen_mackerel_plugin.pl (hunk 5 above)
Comment (Songmu) [inline tool/gen_mackerel_plugin.pl]: Add t/convention.t to check it. I'll document this convention later.

File: tool/gen_mackerel_plugin.pl (hunk 1 above)
Comment (Songmu) [inline tool/gen_mackerel_plugin.pl]: I want to use `strings.Join()`, so stay it Slice.

File: mackerel-plugin.go (hunk 2 above)
Comment (Songmu) [inline mackerel-plugin.go]: I know this problem.
However, I don't expect much to use the `mackerel-plugin` command directly, but to be used via symlink, so it is not so much a problem.

File: mackerel-plugin.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+	err = runPlugin(plug)
+
+	if err != nil {
+		return exitError
+	}
+	return exitOK
+}
+
+var version, gitcommit string
```
Comment (haya14busa) [inline mackerel-plugin.go]: correct me if i'm wrong, but `version` is not set in build as far as I understand and `gitcommit` is unused.

File: mackerel-plugin.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+		plug = strings.TrimPrefix(base, "mackerel-plugin-")
+	} else {
+		if len(args) < 2 {
+			printHelp()
+			return exitError
+		}
+		plug = args[1]
+		if helpReg.MatchString(plug) {
+			printHelp()
+			return exitOK
```
Comment (haya14busa) [inline mackerel-plugin.go]: [nit] I think exit code should be `exitError` here for consistency. (or we can use `2` for printing help)

File: mackerel-plugin.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+		}
+		osargs := []string{f}
+		osargs = append(osargs, args[2:]...)
+		os.Args = osargs
```
Comment (haya14busa) [inline mackerel-plugin.go]: [nit] `os.Args = append([]string{f}, args[2:]...)` should be enough.

**Pull request merged**